binding.cc

```diff
@@ -1269,84 +1259,18 @@ NAPI_METHOD(iterator_init) {
   uint32_t highWaterMark = Uint32Property(env, options, "highWaterMark",
                                           16 * 1024);
 
-  // TODO simplify and refactor the hideous code below
-
-  leveldb::Slice* start = NULL;
-  char *startStr = NULL;
-  CHECK_PROPERTY(start, {
-    start = new leveldb::Slice(_startCh_, _startSz_);
-    startStr = _startCh_;
-  });
-
-  std::string* end = NULL;
-  CHECK_PROPERTY(end, {
-    end = new std::string(_endCh_, _endSz_);
-    delete [] _endCh_;
-  });
-
-  std::string* lt = NULL;
-  CHECK_PROPERTY(lt, {
-    lt = new std::string(_ltCh_, _ltSz_);
-    delete [] _ltCh_;
-    if (reverse) {
-      if (startStr != NULL) {
-        delete [] startStr;
-        startStr = NULL;
-      }
-      if (start != NULL) {
-        delete start;
-      }
-      start = new leveldb::Slice(lt->data(), lt->size());
-    }
-  });
-
-  std::string* lte = NULL;
-  CHECK_PROPERTY(lte, {
-    lte = new std::string(_lteCh_, _lteSz_);
-    delete [] _lteCh_;
-    if (reverse) {
-      if (startStr != NULL) {
-        delete [] startStr;
-        startStr = NULL;
-      }
-      if (start != NULL) {
-        delete start;
-      }
-      start = new leveldb::Slice(lte->data(), lte->size());
-    }
-  });
-
-  std::string* gt = NULL;
-  CHECK_PROPERTY(gt, {
-    gt = new std::string(_gtCh_, _gtSz_);
-    delete [] _gtCh_;
-    if (!reverse) {
-      if (startStr != NULL) {
-        delete [] startStr;
-        startStr = NULL;
-      }
-      if (start != NULL) {
-        delete start;
-      }
-      start = new leveldb::Slice(gt->data(), gt->size());
-    }
-  });
-
-  std::string* gte = NULL;
-  CHECK_PROPERTY(gte, {
-    gte = new std::string(_gteCh_, _gteSz_);
-    delete [] _gteCh_;
-    if (!reverse) {
-      if (startStr != NULL) {
-        delete [] startStr;
-        startStr = NULL;
-      }
-      if (start != NULL) {
-        delete start;
-      }
-      start = new leveldb::Slice(gte->data(), gte->size());
-    }
-  });
+  std::string* start = NULL;
+  std::string* end = RangeOption(env, options, "end");
+  std::string* lt = RangeOption(env, options, "lt");
+  std::string* lte = RangeOption(env, options, "lte");
+  std::string* gt = RangeOption(env, options, "gt");
+  std::string* gte = RangeOption(env, options, "gte");
+
+  if (!reverse && gte != NULL) start = new std::string(*gte);
```

Comment:
Could also do `start = gte` (no copy), seeing as they all have the same lifetime, but I figured it'd be a useless optimization.